frame/contracts/src/wasm/mod.rs

```diff
@@ -18,114 +18,155 @@
 //! This module provides a means for executing contracts
 //! represented in wasm.
 
+#[macro_use]
+mod env_def;
+mod code_cache;
+mod prepare;
+mod runtime;
+
 use crate::{
 	CodeHash, Schedule, Config,
 	wasm::env_def::FunctionImplProvider,
-	exec::Ext,
+	exec::{Ext, Executable, ExportedFunction},
 	gas::GasMeter,
 };
 use sp_std::prelude::*;
 use sp_core::crypto::UncheckedFrom;
 use codec::{Encode, Decode};
-
-#[macro_use]
-mod env_def;
-mod code_cache;
-mod prepare;
-mod runtime;
-
-use self::code_cache::load as load_code;
+use frame_support::dispatch::{DispatchError, DispatchResult};
 use pallet_contracts_primitives::ExecResult;
-
-pub use self::code_cache::save as save_code;
-#[cfg(feature = "runtime-benchmarks")]
-pub use self::code_cache::save_raw as save_code_raw;
 pub use self::runtime::{ReturnCode, Runtime, RuntimeToken};
 
 /// A prepared wasm module ready for execution.
 #[derive(Clone, Encode, Decode)]
-pub struct PrefabWasmModule {
+pub struct PrefabWasmModule<T: Config> {
 	/// Version of the schedule with which the code was instrumented.
 	#[codec(compact)]
 	schedule_version: u32,
+	/// Initial memory size of a contract's sandbox.
 	#[codec(compact)]
 	initial: u32,
+	/// The maximum memory size of a contract's sandbox.
 	#[codec(compact)]
 	maximum: u32,
+	/// The number of alive contracts that use this as their contract code.
+	///
+	/// If this number drops to zero this module is removed from storage.
+	#[codec(compact)]
+	refcount: u64,
```

Comment:
These are the problems that arise when trying to get away from the eager storage model. This is due for a major overhaul like #6439. The code should walk back the call stack and change the refcount in memory. But until then some documentation must suffice.

Reply:
Yes you are right, I thought store were using the cached refcount but they don't, so the outdated reference is not used by `store` and `store_decrement`. Only `occupied_storage` could potentially make use of an outdated refcount (maybe this function `occupied_storage` could have a comment, but all current usage are fine)

Reply:
Added some comments. But I also changed the code so that `execute` consumes `self` and does the storage itself. This way we prevent holding a `Executable` after an execution point that might outdate the refcount.